File: TacTacToe/TacTacToeCS/Computer.cs
namespace TacTacToeCS
{
    class Computer : Player
    {
        bool fMaxPlayer;
        int fNextMove;

        public Computer(bool aMaxPlayer)
        {
            fMaxPlayer = aMaxPlayer;
        }

        public override int MakeMove(Board aBoard)
        {
            Minimax(aBoard, fMaxPlayer);
            return fNextMove;
        }

        int Minimax(Board aBoard, bool aMaxPlayer)
        {
            if (aBoard.GameOver()) return aBoard.WinningPlayer();

            int[] lValidMoves = aBoard.ValidMoves();
            int lBestScore;
            int lBestMove = new int();

            if (aMaxPlayer) lBestScore = int.MinValue;
            else            lBestScore = int.MaxValue;

            for (int i = 0; i < lValidMoves.Length; i++)
            {
                int lValue = Minimax(new Board(aBoard, lValidMoves[i], aMaxPlayer ? 1 : -1), !aMaxPlayer);

                if (aMaxPlayer && lValue > lBestScore || !aMaxPlayer && lValue < lBestScore)
                {
                    lBestScore = lValue;
                    lBestMove = lValidMoves[i];
                }
            }

            fNextMove = lBestMove;
            return lBestScore;
        }
    }
}

File: TacTacToe/TicTacToeAllegro/Computer.cpp
#include "Computer.h"

int Computer::Minimax(Board aBoard, bool aMax)
{
	if (aBoard.GameOver())
	{
		return aBoard.WinningPlayer();
	}

	int lBestScore;

	if (aMax)	lBestScore = -2;
	else		lBestScore = 2;

	for (int i = 0; i < 9; i++)
	{
		if (aBoard.ValidMove(i))
		{
			Board Copy(aBoard);
			Copy.MakeMove(i, aMax ? 1 : -1);

			int lValue = Minimax(Copy, !aMax);

			if ((aMax && lValue > lBestScore) || (!aMax && lValue < lBestScore))
			{
				lBestScore = lValue;
			}
		}
	}

	return lBestScore;
}

int Computer::MakeMove(Board aBoard)
{
	int lBestScore = 1;
	int lBestMove = -1;

	for (int i = 0; i < 9; i++)
	{
		if (aBoard.ValidMove(i))
		{
			Board Copy(aBoard);
			Copy.MakeMove(i, -1);

			int lCurrentScore = Minimax(Copy, true);

			if (lCurrentScore < lBestScore)
			{
				lBestScore = lCurrentScore;
				lBestMove = i;
			}
		}
	}

	return lBestMove;
}